### PR TITLE
ENH: Add the alternating dual update method

### PR DESCRIPTION
**Pull request by sbanert**
This pull request adds a new (nonsmooth) solver, which is, according to the authors of the corresponding paper, especially good in doing tomography problems on a graphics card.

It contains the changes of #1166, and the corresponding test uses the functionality provided by #1166 as requested by Jonas in the discussion there.

### CODE REVIEW

File: odl/solvers/nonsmooth/adupdates.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        else:
+            rng = range(len(ops))
+
+        for i in rng:
+            step = stepsize / majs[i]
+            arg = duals[i] + step * ops[i](x)
+            tmp_ran = tmp_rans[ops[i].range]
+            proxs[i](arg, out=tmp_ran)
+            x -= 1.0 / stepsize * ops[i].adjoint(tmp_ran - duals[i])
+            duals[i] = tmp_ran.copy()
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: duals[i].assign(tmp_ran)

File: odl/solvers/nonsmooth/adupdates.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        # for i in range(length)])
+        for i in range(length):
+            x -= 1.0 / stepsize * ops[i].adjoint(duals[i])
+
+        if random:
+            rng = np.random.permutation(range(len(ops)))
+        else:
+            rng = range(len(ops))
+
+        for i in rng:
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: Both loops use `i` as iteration variable, this is confusing. Prefer using different names, e.g. i and j

File: odl/solvers/nonsmooth/adupdates.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            step = stepsize / majs[i]
+            arg = duals[i] + step * ops[i](x)
+            tmp_ran = tmp_rans[ops[i].range]
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: This whole code creates copies all over the place, I think it would be better to simply remove this temporary for now (alternatively fix all the copies, but that would be some work and I'm not sure how much it will help atm).

File: odl/solvers/nonsmooth/adupdates.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Alternating dual (AD) upgrade algorithm studied by McGaffin and Fessler.
+
+The alternating dual upgrade method is reported to be a flexible method which
+allows an acceleration by permitting efficient calculations on a graphics card.
+"""
+
+# Imports for common Python 2/3 codebase
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/solvers/nonsmooth/adupdates.py]: Yes we removed those comments.

File: odl/solvers/nonsmooth/adupdates.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
+# Copyright 2014-2017 The ODL contributors
+#
+# This file is part of ODL.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Alternating dual (AD) upgrade algorithm studied by McGaffin and Fessler.
```
Comment (kohr-h) [inline odl/solvers/nonsmooth/adupdates.py]: below it says "update" not "upgrade"

File: odl/solvers/nonsmooth/adupdates.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+
+import numpy as np
+
+__all__ = ('adupdates',)
+
+
+def adupdates(funcs, ops, x, stepsize, majs, niter, random=False,
+              callback=None, callback_loop='outer'):
```
Comment (kohr-h) [inline odl/solvers/nonsmooth/adupdates.py]: I suggest that we use the same signature (where possible) as for the Douglas-Rachford method, i.e. `g` for functionals that are composed, `L` for the operators, etc.
See also #1197

File: odl/solvers/nonsmooth/adupdates.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+    * Assume that `f_j` is a `L1_norm` or a `L2_norm_squared`, then setting
+    `majs[i]` a `f_j.domain.element` :math:`z` corresponds to the choice
+    :math:`\mathrm{diag}(z)`.
+
+    References
+    ----------
+    [MF2015] McGaffin, M G, and Fessler, J A. *Alternating dual updates
+    algorithm for X-ray CT reconstruction on the GPU*. IEEE Transactions
+    on Computational Imaging, 1.3 (2015), pp 186--199.
```
Comment (kohr-h) [inline odl/solvers/nonsmooth/adupdates.py]: Reference looks perfect. If you want to use a citation with link, which would be something for the short summary before `Parameters`, you can use a link like
```
`[MF2015] <URL for DOI, arXiv or whatever>`_
```
This is at least my current favorite way of doing it, because it provides a clickable link *and* a text reference in the `References` section.

File: odl/solvers/nonsmooth/adupdates.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            arg = duals[i] + step * ops[i](x)
+            tmp_ran = tmp_rans[ops[i].range]
+            proxs[i](arg, out=tmp_ran)
+            x -= 1.0 / stepsize * ops[i].adjoint(tmp_ran - duals[i])
+            duals[i] = tmp_ran.copy()
+
+            if callback is not None and callback_loop == 'inner':
+                callback(x)
+        if callback is not None and callback_loop == 'outer':
+            callback(x)
```
Comment (kohr-h) [inline odl/solvers/nonsmooth/adupdates.py]: This is a pretty complex optimization method. I'm sure it's a good one.
What I can recommend is to make a "dead simple" implementation of the same optimizer that does not do any fancy stuff like caching, in-place operation of optimized arithmetic. Just with as little code as possible. That's really good for debugging. See for example [here](https://github.com/odlgroup/odl/blob/87e04941f4337de190654897691327f8964c0f03/odl/solvers/nonsmooth/admm.py#L149-L163).

File: odl/solvers/nonsmooth/adupdates.py (hunk 6 above)
Comment (sbanert) [reply odl/solvers/nonsmooth/adupdates.py]: I am currently doing this and I noticed that there is the parameter kwargs, which is not documented in the docstring of the Douglas–Rachford method. In the iterative methods, this parameter is not present at all. Which variant shall I implement here?

Comment (kohr-h) [reply odl/solvers/nonsmooth/adupdates.py]: We don't really have a rule on that, but if you feel that the signature gets a bit too long with all the optional parameters, stuff the less common, less important or more advanced ones into `kwargs`. In this case, I'd say it's fine without `kwargs` since you have only one optional parameter `random` (I don't count `callback`s).

**Comment by sbanert** [comment]
Should be ready from my side if Travis does not complain.

File: odl/solvers/nonsmooth/adupdates.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+
+        min_x sum_i g_i(L_i x)
+
+    where ``g_i`` are proper, convex and lower semicontinuous functions and
+    ``L_i`` are linear `Operator`s.
+
+    Parameters
+    ----------
+    g : sequence of `Functional`s
+        All functions need to provide a `convex_conj` with a
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: dead link here, you need to write `Functional.convex_conj`

File: odl/solvers/nonsmooth/adupdates.py [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+        `proximal` factory.
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: Same as above, `Functional.proximal`

File: odl/solvers/nonsmooth/adupdates.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    L : sequence of `Operator`s
+        Needs to provide as many elements as `g`.
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: This sentence could use some clarification. What is provided? Typically we say that the length of the sequence should be something

File: odl/solvers/nonsmooth/adupdates.py [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+    x : `LinearSpaceElement`
+        Initial point, updated in-place.
+    stepsize : positive float
+        The stepsize for the outer (proximal point) iteration. The theory
+        guarantees convergence for any positive real number, but the
+        performance might depend on the choice of a good stepsize.
+    majs : sequence of majorizers
+        Parameters determining the stepsizes for the inner iterations. Must be
+        matched with the norms of `L`, and convergence is guaranteed if
+        `majs` are large enough. See the Notes section for details.
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: majs is a dead link, use double \`\` for non-link

File: odl/solvers/nonsmooth/adupdates.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+        `proximal` factory.
+    L : sequence of `Operator`s
+        Needs to provide as many elements as `g`.
+    x : `LinearSpaceElement`
+        Initial point, updated in-place.
+    stepsize : positive float
+        The stepsize for the outer (proximal point) iteration. The theory
+        guarantees convergence for any positive real number, but the
+        performance might depend on the choice of a good stepsize.
+    majs : sequence of majorizers
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: majorizers is still not very obvious. What does this mean? If it is functions, just write "sequence of callables"

File: odl/solvers/nonsmooth/adupdates.py [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+        Parameters determining the stepsizes for the inner iterations. Must be
+        matched with the norms of `L`, and convergence is guaranteed if
+        `majs` are large enough. See the Notes section for details.
+    niter : int
+        Number of (outer) iterations.
+    random : bool, optional
+        If `True`, the order of the dual upgdates is chosen randomly,
+        otherwise the order provided by the lists `g`, `L` and `majs`
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: dead links

File: odl/solvers/nonsmooth/adupdates.py [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+        is used.
+    callback : callable, optional
+        Function called with the current iterate after each iteration.
+    callback_loop : {'inner', 'outer'}, optional
+       If 'inner', the `callback` function is called after each inner
+       iteration, i.e., after each dual update. If 'outer', the `callback`
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: dead link

File: odl/solvers/nonsmooth/adupdates.py [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+    .. math::
+        S_n(x; v_1, \ldots, v_m) = \sum_{i=1}^m (\langle v_i, L_i x \rangle
+            - g_i^*(v_i)) + \frac{\mu}{2} \|x - x_n\|^2.
+
+    Given the dual variables, the new primal variable :math:`\tilde x_{n+1}`
+    can be calculated by directly minimizing :math:`S_n` with respect to
+    :math:`x`. This corresponds to the formula
+
+    .. math::
+         \tilde x_{n+1} = x_n - \frac{1}{\mu} \sum_{i=1}^m L_i^* v_i.
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: What does tilde mean here?

File: odl/solvers/nonsmooth/adupdates.py [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+
+    References
+    ----------
+    [MF2015] McGaffin, M G, and Fessler, J A. *Alternating dual updates
+    algorithm for X-ray CT reconstruction on the GPU*. IEEE Transactions
+    on Computational Imaging, 1.3 (2015), pp 186--199.
+    """
+    # Check the lenghts of the lists (= number of dual variables)
+    length = len(g)
+    if not len(L) == length:
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]:
```python
if len(L) != length:
```

is the preferable style

File: odl/solvers/nonsmooth/adupdates.py [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+                         ''.format(len(L), length))
+
+    if not len(majs) == length:
+        raise ValueError('Number of `majs` {} and number of `g` {} do not'
+                         ' agree.'.format(len(majs), length))
+
+    # Check if operators have a common domain
+    # (the space of the primal variable):
+    domain = L[0].domain
+    if any(domain != opi.domain for opi in L):
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: prefer swapping the order here to follow natural language, e..g. you would typically not say "if the expected domain is not equal to the domain", rather you would say "if the domain is not equal to the expected domain".

In code,

```python
if any(opi.domain != domain for opi in L):
```

File: odl/solvers/nonsmooth/adupdates.py [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+    """
+    # Check the lenghts of the lists (= number of dual variables)
+    length = len(g)
+    if not len(L) == length:
+        raise ValueError('Number of `L` {} and number of `g` {} do not agree.'
+                         ''.format(len(L), length))
+
+    if not len(majs) == length:
+        raise ValueError('Number of `majs` {} and number of `g` {} do not'
+                         ' agree.'.format(len(majs), length))
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: We typically write the exception messages in lower case characters and without a dot, e.g.

```python
'number of `majs` {} and number of `g` {} do not agree'
```

See https://github.com/odlgroup/odl/pull/427 if you want a rationalle for this. Basically we follow the python core standard.

File: odl/solvers/nonsmooth/adupdates.py [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+
+The alternating dual upgrade method is reported to be a flexible method which
+allows an acceleration by permitting efficient calculations on a GPU.
+"""
+
+from __future__ import print_function, division, absolute_import
+
+import numpy as np
+
+__all__ = ('adupdates', 'adupdates_simple')
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: The function `adupdates_simple` should not be exported.

File: odl/test/solvers/nonsmooth/adupdates_test.py [hunk 21]
```diff
@@ -0,0 +1,106 @@
+# Copyright 2014-2017 The ODL contributors
+#
+# This file is part of ODL.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Test for the Fessler alternating dual updates solver."""
```
Comment (adler-j) [inline odl/test/solvers/nonsmooth/adupdates_test.py]: Since "Fessler" is not part of the ODL name, it should not be mentioned here either

File: odl/test/solvers/nonsmooth/adupdates_test.py [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+
+from __future__ import division
+# import pytest
```
Comment (adler-j) [inline odl/test/solvers/nonsmooth/adupdates_test.py]: Remove this line

File: odl/test/solvers/nonsmooth/adupdates_test.py [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+import odl
+# import numpy as np
```
Comment (adler-j) [inline odl/test/solvers/nonsmooth/adupdates_test.py]: Remove this line

File: odl/test/solvers/nonsmooth/adupdates_test.py [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+    can be given in closed form. If we set
+
+    (a)   (25/12)                    (x_1) = (1)
+    (b)   (77/60)                    (x_2) = (1)
+    (c) = (19/20)         then       (x_3) = (1)
+    (d)   (319/420),                 (x_4) = (1).
+
+    We therefore solve the problem
+
+    min ||Ax - b||^2 + TV(x) s.t. x >= 0
```
Comment (adler-j) [inline odl/test/solvers/nonsmooth/adupdates_test.py]: I was fine until this point. Why should this have the same solution as above? What does TV even mean here?

File: odl/test/solvers/nonsmooth/adupdates_test.py [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+    mat1 = [[1, 1 / 2, 1 / 3, 1 / 4],
+            [1 / 2, 1 / 3, 1 / 4, 1 / 5]]
+    mat2 = [[1 / 3, 1 / 4, 1 / 5, 1 / 6],
+            [1 / 4, 1 / 5, 1 / 6, 1 / 7]]
+
+    # Create the linear operators
+    mat1op = odl.MatrixOperator(mat1)
+    mat2op = odl.MatrixOperator(mat2)
+    domain = mat1op.domain
+    tv1 = odl.MatrixOperator([[1.0, -1.0, 0.0, 0.0], [0.0, 0.0, 1.0, -1.0]])
```
Comment (adler-j) [inline odl/test/solvers/nonsmooth/adupdates_test.py]: break line here for better readability

File: odl/test/solvers/nonsmooth/adupdates_test.py [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+    nnegm = nneg.range.element([1.0, 1.0, 1.0, 1.0])
+    majs = [mat1m, mat2m, tv1m, tv2m, nnegm]
+
+    expected_solution = domain.element([1, 1, 1, 1])
+    # Create right-hand-sides of the equation
+    rhs1 = mat1op.range.element(mat1op(expected_solution))
+    rhs2 = mat2op.range.element(mat2op(expected_solution))
+
+    # Create the functionals
+    fid1 = odl.solvers.functional.functional.FunctionalTranslation(
```
Comment (adler-j) [inline odl/test/solvers/nonsmooth/adupdates_test.py]: No need to use `odl.solvers.functional.functional.FunctionalTranslation`, simply use `.translated(...)` on the functional

File: odl/test/solvers/nonsmooth/adupdates_test.py [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+    mat1m = mat1op(mat1op.adjoint(mat1op.range.one()))
+    mat2m = mat2op(mat2op.adjoint(mat2op.range.one()))
+    tv1m = tv1.range.element([2.0, 2.0])
+    tv2m = 2.0
+    nnegm = nneg.range.element([1.0, 1.0, 1.0, 1.0])
+    majs = [mat1m, mat2m, tv1m, tv2m, nnegm]
+
+    expected_solution = domain.element([1, 1, 1, 1])
+    # Create right-hand-sides of the equation
+    rhs1 = mat1op.range.element(mat1op(expected_solution))
```
Comment (adler-j) [inline odl/test/solvers/nonsmooth/adupdates_test.py]: `mat1op.range.element` should not be needed here

File: odl/test/solvers/nonsmooth/adupdates_test.py [hunk 28]
```diff
@@ -0,0 +1,106 @@
...
+    rhs2 = mat2op.range.element(mat2op(expected_solution))
+
+    # Create the functionals
+    fid1 = odl.solvers.functional.functional.FunctionalTranslation(
+        odl.solvers.functional.default_functionals.L2NormSquared(
+            mat1op.range), rhs1)
+    fid2 = odl.solvers.functional.functional.FunctionalTranslation(
+        odl.solvers.functional.default_functionals.L2NormSquared(
+            mat2op.range), rhs2)
+    reg1 = odl.solvers.functional.default_functionals.L1Norm(tv1.range)
```
Comment (adler-j) [inline odl/test/solvers/nonsmooth/adupdates_test.py]: odl.solvers.L1Norm(tv1.range) should suffice

File: odl/test/solvers/nonsmooth/adupdates_test.py [hunk 29]
```diff
@@ -0,0 +1,106 @@
...
+    reg2 = odl.solvers.functional.default_functionals.L1Norm(tv2.range)
+    ind = odl.solvers.functional.default_functionals.IndicatorNonnegativity(
+        nneg.range)
+    funcs = [fid1, fid2, reg1, reg2, ind]
+
+    # Start from zero
+    x = tv1.domain.zero()
+    x_simple = tv1.domain.zero()
+
+    # Select stepsize one
```
Comment (adler-j) [inline odl/test/solvers/nonsmooth/adupdates_test.py]: This line is somewhat redundant, it follows from the variable name. Documentation should not document "obvious" stuff. In general, the rule is that documentation should be about "why" something is done. "What" is done should follow from the code.

File: odl/solvers/nonsmooth/adupdates.py [hunk 30]
```diff
@@ -0,0 +1,210 @@
+# Copyright 2014-2017 The ODL contributors
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/solvers/nonsmooth/adupdates.py]: I prefer one sweeping update for just that.

File: odl/solvers/nonsmooth/adupdates.py [hunk 31]
```diff
@@ -0,0 +1,210 @@
...
+__all__ = ('adupdates', 'adupdates_simple')
+
+
+def adupdates(x, g, L, stepsize, majs, niter, random=False,
+              callback=None, callback_loop='outer'):
+    """Alternating Dual updates method.
+
+    The Alternating Dual (AD) updates method of McGaffin and Fessler `[MF2015]
+    <http://ieeexplore.ieee.org/document/7271047/>`_ is designed to solve an
+    optimization problem of the form
```
[the remark this answers is not in the document]
Comment (kohr-h) [reply odl/solvers/nonsmooth/adupdates.py]: We need something here, yes. If the text ends like this
```
text text text::
    code code code
```
the colon will be rendered by Sphinx. If the text should not end with a colon (like here), place a whitespace before the double colon, like this:
```
text text text ::
    code code code
```

File: odl/solvers/nonsmooth/adupdates.py (hunk 16 above)
Comment (kohr-h) [reply odl/solvers/nonsmooth/adupdates.py]: Remark here that applies to a lot of places in our docstrings:

If we use [raw string literals](https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals) whenever there's math in it, we can ditch all those double backslash ugliness.

Here, some math won't work (e.g. `\frac`), so the simplest fix is to just prepend the whole docstring with an `r`.

However, we should explore what happens with unicode stuff, in cases where names require non-ASCII characters.

File: odl/solvers/nonsmooth/adupdates.py (hunk 19 above)
Comment (kohr-h) [reply odl/solvers/nonsmooth/adupdates.py]: Small addition to this: it's more helpful for the user if the expected behavior is explicitly stated, e.g.,
```
'`len(majs)` should be equal `len(g)`, but {} != {}'
```
The library could definitely use a refreshment in that department, but new code should act as a good example 😄

File: odl/solvers/nonsmooth/adupdates.py (hunk 20 above)
Comment (sbanert) [reply odl/solvers/nonsmooth/adupdates.py]: But if I do not export this function, how can I use it in tests or for debugging?

Comment (adler-j) [reply odl/solvers/nonsmooth/adupdates.py]: It is still available under the full path `odl.solvers.nonsmooth.adaupdates.adaupdates_simple`.

File: odl/solvers/nonsmooth/adupdates.py (hunk 9 above)
Comment (sbanert) [reply odl/solvers/nonsmooth/adupdates.py]: I tried this by building the documentation. `Functional.convex_conj` is a dead link, too.

File: odl/solvers/nonsmooth/adupdates.py (hunk 16 above)
Comment (sbanert) [reply odl/solvers/nonsmooth/adupdates.py]: Unicode seems to work fine at least in my environment with Python 3.6.

File: odl/solvers/nonsmooth/adupdates.py (hunk 20 above)
Comment (kohr-h) [reply odl/solvers/nonsmooth/adupdates.py]: Technically, the `__all__` thing is for defining explicitly what should be exposed to wildcard imports, i.e. `from odl.solvers.nonsmooth.adaupdates import *`. If it's undefined, the wildcard import just imports everything not starting with an underscore (roughly).
For us, defining stuff in `__all__` is also a mechanism to make things available on a higher level, since the `__init__.py` files on higher levels do the wildcard import for all submodules (with exceptions).

So it's always possible to import names by going all the way down to where the name is actually defined (= bound to a value).

File: odl/solvers/nonsmooth/adupdates.py (hunk 9 above)
Comment (kohr-h) [reply odl/solvers/nonsmooth/adupdates.py]: I've never really understood the rules as to when you can use the name without full path. @adler-j can you nail it down?

File: odl/solvers/nonsmooth/adupdates.py (hunk 19 above)
Comment (kohr-h) [reply odl/solvers/nonsmooth/adupdates.py]: "equal to" that is

File: odl/solvers/nonsmooth/adupdates.py (hunk 9 above)
Comment (sbanert) [reply odl/solvers/nonsmooth/adupdates.py]: Maybe it can’t find a destination. Links would have to go to entries of the attributes table in this file: https://odlgroup.github.io/odl/generated/odl.solvers.functional.functional.Functional.html

Actually, it would be nice to have the complete docstring of Functional.proximal and Functional.convex_conj printed somewhere in the docs.

**Comment by sbanert** [comment]
I did everything except

* the dead links Functional.convex_conj and Functional.proximal
* upgrading to 2018.

File: odl/solvers/nonsmooth/adupdates.py [hunk 32]
```diff
@@ -0,0 +1,217 @@
...
+The alternating dual upgrade method solves structured convex optimization
+problems by successively updating dual variables which are associated with
+each of the components.
+"""
+
+from __future__ import print_function, division, absolute_import
+
+import numpy as np
+
+__all__ = ('adupdates_method',)
```
Comment (adler-j) [inline odl/solvers/nonsmooth/adupdates.py]: Why is it called "method" btw, wouldn't `adaupdates` look nicer?

File: odl/test/solvers/nonsmooth/adupdates_test.py [hunk 33]
```diff
@@ -0,0 +1,102 @@
...
+    (c) = (19/20)         then       (x_3) = (1)
+    (d)   (319/420),                 (x_4) = (1).
+
+    We solve the problem
+
+    min ||Ax - b||^2 + TV(x) s.t. x >= 0
+
+    for the matrix A, the r.h.s. b as above and the total variation TV, which
+    is given as the sum of the absoulute values of consecutive entries of the
+    solution. The solution of this problem is clearly x = (1, 1, 1, 1), since
```
Comment (adler-j) [inline odl/test/solvers/nonsmooth/adupdates_test.py]: Well this is assuming that we're using periodic boundary conditions for TV (which is not default in ODL), but I guess we can let that pass for now :O

Comment (sbanert) [reply odl/test/solvers/nonsmooth/adupdates_test.py]: Not really, this total variation takes three differences for a four-component vector. This is still not the standard behaviour, though, because in ODL the absolute value of the last element would be added to the functional (which was very irritating to most of my students).

**Comment by sbanert** [comment]
I fixed the issues and corrected a bug. Please do *not* merge, though, because this is a rebase on the branch of #1283 (otherwise #1279 would have been triggered).

**Comment by sbanert** [comment]
Now, I rebased on master, but I think I will have to rebase once more because Travis will probably fail.

**Comment by kohr-h** [comment]
Seems so, but after that it should all be fine.

**Comment by aringh** [comment]
The fix for Travis doc build is now in master, so if you rebase again all test should hopefully pass :-)

**Comment by adler-j** [comment]
Please perform that rebase!

**Comment by sbanert** [comment]
Succeeded.

**Comment by adler-j** [comment]
Merged, well done!
